ray-curator/ray_curator/stages/deduplication/removal.py

```diff
@@ -0,0 +1,102 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Removal stage for distributed deduplication pipeline.
+
+This stage implements the removal phase of the distributed deduplication approach:
+1. Takes a DocumentBatch and determines the min/max ID range
+2. Filters the parquet files for IDs to remove within this range
+3. Filters out documents based on the removal list
+4. Returns the filtered DocumentBatch
+"""
+
+import time
+from dataclasses import dataclass
+from typing import Any
+
+import pandas as pd
+
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.stages.deduplication.id_generator import CURATOR_DEDUP_ID_STR
+from ray_curator.tasks import DocumentBatch
+
+
+@dataclass
+class DuplicatesRemovalStage(ProcessingStage[DocumentBatch, DocumentBatch]):
+    """
+    Stage for removing duplicate documents based on pre-computed removal lists.
+
+    Args:
+        ids_to_remove_path: Path to parquet files containing IDs to remove
+        id_field: Field to use for deduplication. Defaults to CURATOR_DEDUP_ID_STR.
+        read_kwargs: Additional arguments for reading parquet files
+    """
+
+    ids_to_remove_path: str
+    id_field: str = CURATOR_DEDUP_ID_STR
+
+    # Optional parameters
+    read_kwargs: dict[str, Any] | None = None
+
+    def __post_init__(self):
+        """Initialize parent class after dataclass initialization."""
+        super().__init__()
+        self._name = "DuplicatesRemovalStage"
+        self.read_kwargs = self.read_kwargs if self.read_kwargs is not None else {}
+
+    def process(self, task: DocumentBatch) -> DocumentBatch:
+        """
+        Our deduplicator should've written out a parquet file with the IDs to remove.
+        We read that file, filter the input dataframe to only include the IDs to remove,
+        and return the filtered dataframe.
+        We optimize by not loading the whole ids to remove into memory, but only loading the ids that are in the range of the input dataframe.
+        """
+        df = task.to_pandas()
+        input_df_t0 = time.perf_counter()
+        min_id = df[self.id_field].min()
+        max_id = df[self.id_field].max()
+        input_df_time = time.perf_counter() - input_df_t0
+        # Filter the parquet files for IDs to remove within this range
+        read_dupes_t0 = time.perf_counter()
+        removal_df = pd.read_parquet(
+            self.ids_to_remove_path,
+            filters=[("id", ">=", min_id), ("id", "<=", max_id)],
```

Comment:
Maybe "id" should not be hardcoded here?

Reply:
I think it makes sense, I'll add id field now